flocker/node/agents/gce.py

```diff
@@ -448,44 +488,58 @@ def create_volume(self, dataset_id, size):
         )
 
     def attach_volume(self, blockdevice_id, attach_to):
-        config = dict(
-            deviceName=blockdevice_id,
-            autoDelete=False,
-            boot=False,
-            source=(
-                "https://www.googleapis.com/compute/v1/projects/%s/zones/%s/"
-                "disks/%s" % (self._project, self._zone, blockdevice_id)
+        with start_action(
+                action_type=u"flocker:node:agents:gce:attach_volume",
+                blockdevice_id=blockdevice_id,
+                attach_to=attach_to,
+        ) as action:
+            config = dict(
+                deviceName=blockdevice_id,
+                autoDelete=False,
+                boot=False,
+                source=(
+                    "https://www.googleapis.com/compute/v1/projects/%s/zones/"
+                    "%s/disks/%s" % (self._project, self._zone, blockdevice_id)
+                )
             )
-        )
-        try:
-            # TODO(mewert): Verify timeout and error conditions.
-            # TODO(mewert): Test what happens when disk is attached RW to a
-            #               different instance, raise the correct error.
-            result = self._do_blocking_operation(
-                self._compute.instances().attachDisk,
-                instance=attach_to,
-                body=config
+            try:
+                # TODO(mewert): Verify timeout and error conditions.
+                # TODO(mewert): Test what happens when disk is attached RW to a
+                #               different instance, raise the correct error.
+                result = self._do_blocking_operation(
+                    self._compute.instances().attachDisk,
+                    instance=attach_to,
+                    body=config
+                )
+            except HttpError as e:
+                if e.resp.status == 400:
+                    # TODO(mewert): verify with the rest API that this is the
+                    # only way to get a 400.
+                    raise UnknownVolume(blockdevice_id)
+                else:
+                    raise e
+            errors = result.get('error', {}).get('errors', [])
+            for e in errors:
+                if e.get('code') == u"RESOURCE_IN_USE_BY_ANOTHER_RESOURCE":
+                    raise AlreadyAttachedVolume(blockdevice_id)
```

Comment:
The HTTP error code is confusingly part of the Rest API. It is possible to start an operation, get a 200 response, and keep querying the status of the operation always getting 200 responses, and have the body of one of those responses look like above.

Click "Equivalent REST" on https://console.cloud.google.com/compute/operationsDetail/zones/us-central1-f/operations/operation-1456191905224-52c66193c5541-87cb8805-b0cc7b7c?project=clusterhq-acceptance for an example.

For some reason they decided to use HTTP error codes to describe the result of operations, despite the fact that long Operations are actually GCE REST Nouns, and should probably have some different status states that are not HTTP statuses.